File: find_duplicate/find_duplicate.py
def findDuplicate(arr:list, n:int):
    # Write your code here.
    # Returns an integer.
    fast = arr[0]
    slow = arr[0]
    
    fast = arr[arr[fast]]
    slow = arr[slow]
    while slow != fast:
        fast = arr[arr[fast]]
        slow = arr[slow]
        
    fast = arr[0]
    while slow != fast:
        fast = arr[fast]
        slow = arr[slow]

    return fast

if __name__ == '__main__':
    arr = [1,2,3, 5, 5, 4]
    print(findDuplicate(arr, 6))

File: find_duplicate/find_duplicate.cpp
#include<bits/stdc++.h>

using namespace std;

int findDuplicate_usingFrequency(vector<int> &arr, int n){
	// Write your code here.
    int freq[n + 1] = {
    0
  };
  for (int i = 0; i < n; i++) {
    if (freq[arr[i]] == 0) {
      freq[arr[i]] += 1;
    } else {
      return arr[i];
    }
  }
  return 0;
}

int findDuplicate_tortoise(vector<int>& nums) {
    int fast = nums[0];
    int slow = nums[0];
    do{
        fast = nums[nums[fast]];
        slow = nums[slow];
    }while(slow!=fast);
    fast = nums[0];
    while(slow!=fast){
        fast = nums[fast];
        slow = nums[slow];
    }
    return fast;
}

int main(){
    vector<int> arr{1,2,4,4,3};
    cout<<findDuplicate_tortoise(arr);
}